proposals/p1787.md

```diff
@@ -0,0 +1,97 @@
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1787)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+The status quo syntax for an expression representing the type of an array has
+some slight annoyances and inconsistencies. This is certainly not a major issue,
+but it does impact users broadly; anyone writing multidimensional arrays.
+
+Specifically, the problem is that the array extents are in reverse order from
+how the array is indexed. That is, given an array of length 5, each element of
+which is an array of 10 `i32`s, consider a variable declaration along with an
+indexing into the "last" element.
+
+```
+var array: [[i32; 10]; 5]
+var element: array[4][9]
+```
+
+The concern is that users would be tempted to write the expression
+`array[9][4]`. More generally, the concern is that users writing multi-indexing
+will be tempted to assume the first index corresponds to the lexically first
+extent in the type when the proper correspondence is to the outermost (and
+hence, with the current syntax, lexically last) extent.
+
+## Background
+
+No background knowledge is required.
+
+## Proposal
+
+This concern can be addressed by simply reversing the order of the type and the
+extent syntactically so that the length appears first. Then the example above
+would be rewritten as
+
+```
+var array: [5; [10; i32]]
+var element: array[4][9]
+```
+
+## Details
+
+Because array indexing is read left-to-right, and each index peels off one layer
+of a multidimensional array, reading indexes left-to-right corresponds to
+indexing from the outermost to innermost extents. By listing the extents before
+the type syntactically, this outer-to-inner ordering becomes compatible with the
+left-to-right ordering.
+
+## Rationale
+
+This proposal addresses Carbon's goal to promote
+[Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write).
+It does not have any affect (positive or negative) on any of the other goals.
+
+## Alternatives considered
+
+It's worth looking at what other languages do for arrays.
+
+**Rust** uses the syntax that is currently Carbon's status quo and therefore
+retains the same problem.
+
+**C/C++** simply invert any sequence of nested array specfiers. An array of five
+elements each of type `int[10]` would be written as `int[5][10]`. This makes
+some sense with the "declaration models usage" legacy, but does not make sense
+in a language such as Carbon which has types use the same expression grammar as
+values.
+
+**Go** also places array indices on the right but with a different syntax.
```

Comment:
```suggestion
**Go** also places array bounds on the left but with a different syntax.
```
(I think all the languages under consideration put array indexes on the right; what distinguishes Go is where the array bounds are written.)